### PR TITLE
Use SymEngine in Physics/Hep

### PR DESCRIPTION
**Pull request by ShikharJ**
TODO:
- [ ] Introduce `S` in `physics/hep/gamma_matrices.py`
```
============================= test process starts ==============================
_____ sympy/physics/hep/tests/test_gamma_matrices.py:test_kahane_algorithm _____
  File "/home/travis/miniconda/envs/test-environment/lib/python3.5/site-packages/sympy/physics/hep/tests/test_gamma_matrices.py", line 190, in test_kahane_algorithm
    execute_gamma_simplify_tests_for_function(tfunc, D=4)
  File "/home/travis/miniconda/envs/test-environment/lib/python3.5/site-packages/sympy/physics/hep/tests/test_gamma_matrices.py", line 39, in execute_gamma_simplify_tests_for_function
    assert _is_tensor_eq(tfunc(t), -4*G(mu11)*G(mu31)*G(mu21) - 4*G(mu31)*G(mu11)*G(mu21))
  File "/home/travis/miniconda/envs/test-environment/lib/python3.5/site-packages/sympy/physics/hep/tests/test_gamma_matrices.py", line 188, in tfunc
    return _simplify_single_line(e)
  File "/home/travis/miniconda/envs/test-environment/lib/python3.5/site-packages/sympy/physics/hep/gamma_matrices.py", line 216, in _simplify_single_line
    t1 = kahane_simplify(t1)
  File "/home/travis/miniconda/envs/test-environment/lib/python3.5/site-packages/sympy/physics/hep/gamma_matrices.py", line 704, in kahane_simplify
    resulting_expr += temp_expr
  File "/home/travis/miniconda/envs/test-environment/lib/python3.5/site-packages/sympy/tensor/tensor.py", line 3860, in __radd__
    return TensAdd(other, self)
  File "/home/travis/miniconda/envs/test-environment/lib/python3.5/site-packages/sympy/tensor/tensor.py", line 2781, in __new__
    TensAdd._tensAdd_check(args)
  File "/home/travis/miniconda/envs/test-environment/lib/python3.5/site-packages/sympy/tensor/tensor.py", line 2858, in _tensAdd_check
    raise ValueError('all tensors must have the same indices')
ValueError: all tensors must have the same indices
```

### CODE REVIEW

**Comment by ShikharJ** [comment]
@isuruf I've tried a number of workarounds for importing `S` but none of them seem to fix the mentioned errors. Can you take a look here and suggest a further course of action?

**Comment by jksuom** [comment]
In SymPy, `S` is the name of the singleton registry. That is its principal role. It also doubles as an abbreviation of `sympify` via the line `__call__ = staticmethod(sympify)`, but that is tricky as it depends on Python's way of parsing parentheses. To avoid confusion, it may sometimes be better to keep these roles separate and import `S` when the registry is required (e.g., `S.One`) and `sympify` when it is really needed (and then use `sympify(...)` instead of `S(...)`).

**Comment by ShikharJ** [comment]
@isuruf @jksuom In that case, can you tell me how to benchmark the changes?

**Comment by ShikharJ** [comment]
@isuruf Do you have any suggestions as to what can be done here?

File: sympy/physics/hep/gamma_matrices.py [hunk 1]
```diff
@@ -25,7 +25,8 @@
     metric(LorentzIndex,LorentzIndex)
 
 """
-from sympy import S, Mul, eye, trace
+from sympy.core.backend import Mul, eye
+from sympy import trace, S
```
Comment (isuruf) [inline sympy/physics/hep/gamma_matrices.py]: You'll need to use symengine for this as well.

File: sympy/physics/hep/tests/test_gamma_matrices.py [hunk 2]
```diff
@@ -214,43 +212,43 @@ def test_kahane_simplify1():
     assert r.equals(t)
     t = G(i0)*G(-i0)
     r = kahane_simplify(t)
-    assert r.equals(4*eye(4))
```
Comment (isuruf) [inline sympy/physics/hep/tests/test_gamma_matrices.py]: This is fine for now, but `.equals` should be implemented in symengine.

File: sympy/physics/hep/gamma_matrices.py (hunk 1 above)
Comment (ShikharJ) [inline sympy/physics/hep/gamma_matrices.py]: I meant about that only. The only way that occurs to me is to implement the `Singleton` pattern in `SymEngine`. Is there a workaround for this?

**Comment by ShikharJ** [comment]
@isuruf Can you tell me how can I checkout the `SymEngine.py` `dev` branch, now that most of the work is in?

**Comment by ShikharJ** [comment]
Same as [here](https://github.com/sympy/sympy/pull/13214#issuecomment-533908620).

**Pull request closed without merging**